plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/UniqueDeclarationChecker.kt

```diff
@@ -0,0 +1,83 @@
+import org.jetbrains.kotlin.fir.symbols.SymbolInternals
+import org.jetbrains.kotlin.name.ClassId
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+
+sealed class UniquenessLevel
+data object Unique : UniquenessLevel()
+data object Shared : UniquenessLevel()
+
+class UniqueDeclarationChecker(private val session: FirSession) : FirSimpleFunctionChecker(MppCheckerKind.Common) {
+
+    private fun getAnnotationId(name: String): ClassId =
+        ClassId(FqName.fromSegments(listOf("org", "jetbrains", "kotlin", "formver", "plugin")), Name.identifier(name))
+
+    private val uniqueId: ClassId = getAnnotationId("Unique")
+
+    private fun getUniqueLevelFromExpression(expression: FirExpression, context: Map<FirVariable, UniquenessLevel>): UniquenessLevel {
+        return Shared
+    }
+
+    @OptIn(SymbolInternals::class)
+    override fun check(declaration: FirSimpleFunction, context: CheckerContext, reporter: DiagnosticReporter) {
+        val errorCollector = ErrorCollector()
+        try {
+            // uniquenessContext is a map from variable to its uniqueness level
+            val uniquenessContext: MutableMap<FirVariable, UniquenessLevel> = mutableMapOf()
+            // iterate through all parameters and checker their annotation
+            declaration.valueParameters.forEach { parameter ->
+                if (parameter.hasAnnotation(uniqueId, session)) {
+                    uniquenessContext[parameter] = Unique
+                } else {
+                    uniquenessContext[parameter] = Shared
+                }
+            }
+
+            // analyze the function body
+            val body = declaration.body ?: return
+            body.statements.forEach { statement: FirStatement ->
+                // check if statement is a function all, otherwise, continue
+                if (statement is FirFunctionCall) {
+                    val functionReference = statement.calleeReference as FirResolvedNamedReference
+                    val functionDeclaration = functionReference.resolvedSymbol.fir as FirSimpleFunction
+                    val functionParameters = functionDeclaration.valueParameters
+                    val functionArguments = statement.arguments
+                    assert(functionArguments.size == functionParameters.size)
+                    // TODO: Should do unify here
+                    for (i in functionParameters.indices) {
+                        val parameter = functionParameters[i]
+                        if (!parameter.hasAnnotation(uniqueId, session)) continue
+                        val argument = functionArguments[i]
+                        if (getUniqueLevelFromExpression(argument, uniquenessContext) == Shared) {
+                            reporter.reportOn(argument.source, PluginErrors.UNIQUENESS_VIOLATION, parameter.name.asString(), context)
+                        }
+                    }
+                }
```

Comment:
I know I said we don't want to write an elaborate visitor, but I think splitting this out into a visitor is still the way to go here; it'll make it much easier to add elements.